File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/SimpleDirectionContainer.java
package co.electric.snake.strategy.bonostrategy;

import co.electric.snake.framework.model.Direction;

import java.util.*;

public class SimpleDirectionContainer extends HashSet<Direction> {

    private static final Random RANDOM = new Random();

    public SimpleDirectionContainer() {
        super();
    }

    public SimpleDirectionContainer(Collection<? extends Direction> collection) {
        super(collection);
    }

    public SimpleDirectionContainer getElementsInANewInstance() {
        return new SimpleDirectionContainer(this);
    }

    public Direction getRandomElement() {
        final List<Direction> elements = new ArrayList<>(this);
        return !elements.isEmpty() ? elements.get(RANDOM.nextInt(elements.size())) : null;
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/SimpleDirectionContainer.kt
package co.electric.snake.strategy.bonostrategy

import co.electric.snake.framework.model.Direction
import java.util.*

class SimpleDirectionContainer : HashSet<Direction> {

    companion object {
        private val RANDOM = Random()
    }

    constructor() : super()
    constructor(collection: Collection<Direction>) : super(collection)

    fun getElementsInANewInstance(): SimpleDirectionContainer {
        return SimpleDirectionContainer(this)
    }

    fun getRandomElement(): Optional<Direction> {
        return Optional.ofNullable(elementAtOrNull(if (size > 0) RANDOM.nextInt(size) else 0))
    }

}
